scipy/optimize/_direct.py

```diff
@@ -0,0 +1,275 @@
+    bounds: Union[Iterable, Bounds],
+    *,
+    args: tuple = (),
+    disp: bool = False,
+    eps: float = 1e-4,
+    maxfun: int = 20000,
+    maxiter: int = 6000,
+    locally_biased: bool = True,
+    f_min: float = -np.inf,
+    f_min_rtol: float = 1e-4,
+    vol_tol: float = 1e-16,
+    len_tol: float = 1e-8,
+    callback: Optional[Callable[[npt.ArrayLike], NoneType]] = None
+) -> OptimizeResult:
+    """
+    Finds the global minimum of a function using the
+    DIRECT algorithm.
+
+    The algorithm is due to Jones et al. [1]_.
+
+    Parameters
+    ----------
+    func : callable
+        The objective function to be minimized.
+        ``func(x, *args) -> float``
+        where ``x`` is an 1-D array with shape (n,) and ``args`` is a tuple of
+        the fixed parameters needed to completely specify the function.
+    bounds : sequence or `Bounds`
+        Bounds for variables. There are two ways to specify the bounds:
+
+        1. Instance of `Bounds` class.
+        2. ``(min, max)`` pairs for each element in ``x``.
+
+    args : tuple, optional
+        Any additional fixed parameters needed to
+        completely specify the objective function.
+    disp : bool, optional
+        If ``True``, print logging information about the optimization process.
+    eps : float, optional
+        Minimal required difference of the objective function values
+        between the current best hyperrectangle and the next potentially
+        optimal hyperrectangle to be divided. In consequence, `eps` serves as a
+        tradeoff between local and global search: the smaller, the more local
+        the search becomes. Default is 1e-4.
+    maxfun : int, optional
+        Approximate upper bound on objective function evaluations.
+        Default is 20000.
+    maxiter : int, optional
+        Maximum number of iterations. Default is 6000.
+    locally_biased : bool, optional
+        If `True` (default), use the locally biased variant [2]_ of the
+        algorithm known as DIRECT_L. If `False`, use the original unbiased
+        DIRECT algorithm [1]_. For hard problems with many local minima,
+        `False` is recommended.
+    f_min : float, optional
+        Function value of the global optimum. Set this value only if the
+        global optimum is known. Default is ``-np.inf``, so that this
+        termination criterion is deactivated.
+    f_min_rtol : float, optional
+        Terminate the optimization once the relative error between the
+        current best minimum `f` and the supplied global minimum `f_min`
+        is smaller than `f_min_rtol`. This parameter is only used if
+        `f_min` is also set. Default is 1e-4.
+    vol_tol : float, optional
+        Terminate the optimization once the volume of the hyperrectangle
+        containing the lowest function value is smaller than `vol_tol`
+        of the complete search space. Must lie between 0 and 1.
+        Default is 1e-16.
```

Comment:
I still think we should loosen the tolerance to make this a bit more competitive by default.

Reply:
I performed another benchmark with the two termination parameters below. 
Suggestion: vol_tol=1e-10, len_tol=1e-5 ?

I find these hard to tune. Maybe in a follow-up PR a termination criterion based on the objective value could be included but updating that C Code will get ugly.


![DIRECT_L_Param_Testing](https://user-images.githubusercontent.com/40656107/159159297-0fc0a637-aa7f-4acd-b2f9-04a283bfc861.png)

Reply:
Not sure I am reading this correctly, but the case in cyan seems to be clearly better overall.

Reply:
Ok I think my description was not clear enough. The graphs show the number of evaluations and elapsed time for DIRECT for the different parameters. The number of solved problems (criterion: global minimum found within relative accuracy of 1e-4) is shown in the legend.

cyan only solves 134 of the benchmark problems compared to 156 for the red curve.

Reply:
Mmm ok the axis mislead me.

> One thing to consider is that vol_tol is sensitive to the dimensionality of the problem: for a 6-dimensional problem a tolerance of 1e-6 would already be reached if the hyperrectangle would cover 10% of the feasible space in each dimension.

Yes these parameters depends on the dimensionality. Unless there are studies on how to fix the parameters with respect to the dimension of the problem, there is not much we can do other than explaining in the doc how to adjust the parameters. It's a good thing to talk about curse of dimensions here as we explicitly talk about hyperrectangles.

Reply:
Added the note. Can be marked as addressed then? :)

Reply:
Thank you, the note is on point 😃 Didn't you say we could change the defaults after your experiments no?

Reply:
Will do so if you agree with vol_tol=1e-10, len_tol=1e-5. Didnt see a definitive conclusion there ;) .

Reply:
You played with this extensively now, so I would trust your judgment.

Reply:
Done. Will update the tests later this week.

Reply:
Tests updated and passing. Cannot mark as resolved for some reason (maybe becauseI am not owner of the repo the PR is from).

Reply:
Resolving this. Thanks. :)